perllib/FixMyStreet/App/Controller/Admin/Users.pm

```diff
@@ -142,31 +142,60 @@ sub add : Local : Args(0) {
         $phone = $parsed_phone if $parsed_phone;
     }
 
-    my $existing_email = $email_v && $c->model('DB::User')->find( { email => $email } );
-    my $existing_phone = $phone_v && $c->model('DB::User')->find( { phone => $phone } );
-    if ($existing_email || $existing_phone) {
-        $c->stash->{field_errors}->{username} = _('User already exists');
-    }
-
-    my $user = $c->model('DB::User')->new( {
-        name => $c->get_param('name'),
-        email => $email ? $email : undef,
-        email_verified => $email && $email_v ? 1 : 0,
-        phone => $phone || undef,
-        phone_verified => $phone && $phone_v ? 1 : 0,
-        from_body => $c->get_param('body') || undef,
-        flagged => $c->get_param('flagged') || 0,
-        # Only superusers can create superusers
-        is_superuser => ( $c->user->is_superuser && $c->get_param('is_superuser') ) || 0,
-    } );
-    $c->stash->{user} = $user;
+    # Previously, an admin could add a staff user without a body (this is
+    # because the 'Staff' checkbox in the form could remain unchecked). A
+    # staff user without a body does not appear in the /admin/users list for
+    # a given cobrand, which has caused confusion for admin, who may try to
+    # create the user again, only to be informed the user already exists.
+    #
+    # To solve this, a) a 'body' parameter is now guaranteed to be passed
+    # through to this code path, and b) if an existing user is submitted, we
+    # check if they have a body; if not, we assign a body to that user so they
+    # will be visible to admin on /admin/users. (If the user already has a
+    # body, we simply return an existence error.)
+    my $from_body = $c->get_param('body') || undef;
+
+    my $user_from_email = $email_v && $c->model('DB::User')->find( { email => $email } );
+    my $user_from_phone = $phone_v && $c->model('DB::User')->find( { phone => $phone } );
+    my $user = $user_from_email || $user_from_phone;
+
+    if ($user) {
+        if ($user->from_body) {
+            $c->stash->{field_errors}->{username} = _('User already exists');
+        }
+
+        $c->stash->{user} = $user;
 
-    return if %{$c->stash->{field_errors}};
+        return if %{$c->stash->{field_errors}};
 
-    $c->forward('user_cobrand_extra_fields');
-    $user->insert;
+        # Assign body to user
+        $user->from_body($from_body);
+        $user->update;
 
-    $c->forward( '/admin/log_edit', [ $user->id, 'user', 'add' ] );
+        $c->forward( '/admin/log_edit', [ $user->id, 'user', 'edit' ] );
```

Comment:
This log line is in both the if and else, so could be moved after it outside.

Reply:
One logs 'edit', the other 'add', so I think I should keep them as is.

Reply:
Though I have rewritten the code to be less repetitive.